File: tests/types_tutorial/stage4_grouped/payments.ts
declare const brand: unique symbol;
type Branded<T, B> = T & { readonly [brand]: B };

export type PaymentId = Branded<string, "PaymentId">;
export type CustomerId = Branded<string, "CustomerId">;
export type VendorId = Branded<string, "VendorId">;

export const RETRY_LIMIT: number = 3;

export type Provider = "stripe" | "adyen" | "wise";
export type PaymentState = "authorized" | "captured" | "refunded" | "failed";
export type Kyc = "unverified" | "pending" | "verified";
export type Payouts = "disabled" | "enabled";

export interface Payment {
  id: PaymentId;
  provider: Provider;
  amount: number;
  state: PaymentState;
  capturedAt?: number;
}

export interface Customer {
  id: CustomerId;
  kyc: Kyc;
}

export interface Vendor {
  id: VendorId;
  payouts: Payouts;
}

export interface Outcome {
  readonly ok: boolean;
  readonly reason?: string;
}

const PROVIDER_VOCABULARY: Record<Provider, Record<string, PaymentState>> = {
  stripe: {
    requires_capture: "authorized",
    succeeded: "captured",
    canceled: "failed",
  },
  adyen: {
    Authorised: "authorized",
    SettleScheduled: "captured",
    Refused: "failed",
  },
  wise: {
    pending: "authorized",
    completed: "captured",
    failed: "failed",
  },
};

export function providerState(
  provider: Provider,
  raw: string,
): PaymentState | undefined {
  return PROVIDER_VOCABULARY[provider][raw];
}

export function findPayment(
  payments: Record<PaymentId, Payment>,
  paymentId: PaymentId,
): Payment | undefined {
  return payments[paymentId];
}

export function authorize(payment: Payment, customer: Customer): Outcome {
  if (payment.amount > 100000 && customer.kyc !== "verified") {
    return { ok: false, reason: "verification required" };
  }
  payment.state = "authorized";
  return { ok: true };
}

export function capture(payment: Payment, vendor: Vendor): Outcome {
  if (payment.state !== "authorized") {
    return { ok: false, reason: "not authorized" };
  }
  if (vendor.payouts !== "enabled") {
    return { ok: false, reason: "vendor cannot receive payouts" };
  }
  payment.state = "captured";
  payment.capturedAt = 1700000000;
  return { ok: true };
}

export function refund(payment: Payment, amount: number): Outcome {
  if (payment.state !== "captured") {
    return { ok: false, reason: "not captured" };
  }
  if (amount > payment.amount) {
    return { ok: false, reason: "refund exceeds capture" };
  }
  payment.state = "refunded";
  return { ok: true };
}

File: tests/types_tutorial/stage4_grouped/payments.py
from dataclasses import dataclass
from enum import StrEnum
from typing import NewType

PaymentId = NewType("PaymentId", str)
CustomerId = NewType("CustomerId", str)
VendorId = NewType("VendorId", str)

RETRY_LIMIT: int = 3

class Provider(StrEnum):
    STRIPE = "stripe"
    ADYEN = "adyen"
    WISE = "wise"

class PaymentState(StrEnum):
    AUTHORIZED = "authorized"
    CAPTURED = "captured"
    REFUNDED = "refunded"
    FAILED = "failed"

class Kyc(StrEnum):
    UNVERIFIED = "unverified"
    PENDING = "pending"
    VERIFIED = "verified"

class Payouts(StrEnum):
    DISABLED = "disabled"
    ENABLED = "enabled"

@dataclass
class Payment:
    id: PaymentId
    provider: Provider
    amount: int
    state: PaymentState
    captured_at: int | None = None

@dataclass
class Customer:
    id: CustomerId
    kyc: Kyc

@dataclass
class Vendor:
    id: VendorId
    payouts: Payouts

@dataclass(frozen=True)
class Outcome:
    ok: bool
    reason: str = ""

PROVIDER_VOCABULARY: dict[Provider, dict[str, PaymentState]] = {
    Provider.STRIPE: {
        "requires_capture": PaymentState.AUTHORIZED,
        "succeeded": PaymentState.CAPTURED,
        "canceled": PaymentState.FAILED,
    },
    Provider.ADYEN: {
        "Authorised": PaymentState.AUTHORIZED,
        "SettleScheduled": PaymentState.CAPTURED,
        "Refused": PaymentState.FAILED,
    },
    Provider.WISE: {
        "pending": PaymentState.AUTHORIZED,
        "completed": PaymentState.CAPTURED,
        "failed": PaymentState.FAILED,
    },
}

def provider_state(provider: Provider, raw: str) -> PaymentState | None:
    return PROVIDER_VOCABULARY[provider].get(raw)

def find_payment(payments: dict[PaymentId, Payment], payment_id: PaymentId) -> Payment | None:
    return payments.get(payment_id)

def authorize(payment: Payment, customer: Customer) -> Outcome:
    if payment.amount > 100000 and customer.kyc != Kyc.VERIFIED:
        return Outcome(False, "verification required")
    payment.state = PaymentState.AUTHORIZED
    return Outcome(True)

def capture(payment: Payment, vendor: Vendor) -> Outcome:
    if payment.state != PaymentState.AUTHORIZED:
        return Outcome(False, "not authorized")
    if vendor.payouts != Payouts.ENABLED:
        return Outcome(False, "vendor cannot receive payouts")
    payment.state = PaymentState.CAPTURED
    payment.captured_at = 1700000000
    return Outcome(True)

def refund(payment: Payment, amount: int) -> Outcome:
    if payment.state != PaymentState.CAPTURED:
        return Outcome(False, "not captured")
    if amount > payment.amount:
        return Outcome(False, "refund exceeds capture")
    payment.state = PaymentState.REFUNDED
    return Outcome(True)
